### PR TITLE
Feature/14 sound acquisition

### LINKED ISSUE #14: Acquisition du son produit par l’instrument

### PR DESCRIPTION
**Pull request by MelvinCou**
Closes #14

### CODE REVIEW

File: assets/utilities/sound/notes.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
+export const notes: { note: string; frequency: number; power: number | undefined }[] = [
```
Comment (MisterPeModder) [inline assets/utilities/sound/notes.ts:1]: Maybe defining a type here would make to code easier to use down the line?
```typescript
export interface Note {
    note: string;
    frequency: number;
    power?: number;
}

// using it
export const notes: Note[] = [
    // ...
];
```

File: assets/utilities/sound/microphone.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
+/**
+ * This code is the retranscription in typescript based on https://github.com/aerik/aerik.github.io
+ * No licence
+ */
```
Comment (MisterPeModder) [inline assets/utilities/sound/microphone.ts:4]: :eyes:

**Review by MisterPeModder — CHANGES_REQUESTED** [review]

File: assets/utilities/sound/microphone.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            micSource.connect(gainNode);
+            micSource.connect(analyser);
+
+            getTones(EnableCanvas);
+        })
+        .catch((err) => console.error(err)); // Disable error in console when blocking mic
+};
+
+//this basically lumps loud tones together and gets their avg frequency
+const getTones = (EnableCanvas: boolean = false) => {
```
Comment (MisterPeModder) [inline assets/utilities/sound/microphone.ts:47]: Why is `EnableCanvas` capitalized?

File: assets/utilities/sound/microphone.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+let analyser: AnalyserNode;
+
+let gainNode: GainNode;
+let hertzBinSize: number;
+let frequencyData: Uint8Array;
+let buflen: number;
+
+const maxValue: number = 256; //based on Uint8Array possible values
+
+const InitMicrophone = (EnableCanvas: boolean = false) => {
```
Comment (MisterPeModder) [inline assets/utilities/sound/microphone.ts:16]: Why is `EnableCanvas` capitalized?

**Review by MisterPeModder — CHANGES_REQUESTED** [review]

File: front/components/HomeLink.vue [hunk 5]
```diff
@@ -16,7 +16,7 @@ const emit = defineEmits<{
 </script>
 
 <template>
-    <div @mouseover="emit('select')" @focusin="emit('select')" class="relative">
+    <div class="relative" @mouseover="emit('select')" @focusin="emit('select')">
```
Comment (MisterPeModder) [inline front/components/HomeLink.vue:19]: Why the change?

File: front/utilities/sound/canvas.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    //horizontal bars
+    for (let n = 0; n < notes.length; n++) {
+        const colString = 'hsl(' + (360 * n) / notes.length + ',100%,80%)';
+        ctx.fillStyle = colString;
+        ctx.fillRect(120, n * 10, notes[n].power ?? 0, -10);
+    }
+    ctx.restore();
+};
+
+export default showCanvas;
```
Comment (MisterPeModder) [inline front/utilities/sound/canvas.ts:33]: This could have been a Vue component

File: front/utilities/sound/microphone.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                stream.getAudioTracks().forEach((track: MediaStreamTrack) => track.stop());
+                // display notes played
+                console.log(notesPlayed);
+            }, trackDuration * 1000);
+        })
+        .catch((err: Error) => console.error(err));
+};
+
+//this basically lumps loud tones together and gets their avg frequency
+const getTones = (enableCanvas = false, DecibelMin: number) => {
```
Comment (MisterPeModder) [inline front/utilities/sound/microphone.ts:66]: `DecibelMin` should be written as `decibelMin`.

File: front/utilities/sound/microphone.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            if (nPtr < notes.length - 2) {
+                nPtr++;
+            }
+        }
+    }
+    if (enableCanvas) {
+        showCanvas(notes);
+    }
+    setTimeout(() => {
+        requestAnimationFrame(() => getTones(enableCanvas, DecibelMin));
```
Comment (MisterPeModder) [inline front/utilities/sound/microphone.ts:170]: This spins indefinitely, completely ignoring the browser's animation framerate.

File: front/views/TestMicView.vue [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<script lang="ts" setup>
+import PlayMic from '@/components/sound/PlayMic.vue';
```
Comment (MisterPeModder) [inline front/views/TestMicView.vue:2]: This file is redundant, please remove it.

**Review by MisterPeModder — CHANGES_REQUESTED** [review]
:clap: :clap:

File: front/utilities/sound/microphone.ts [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+let maxBytePlayed = 0; // max gain of note played
+let startTimeStamp = 0;
+// canvas
+const maxValue = 256; //based on Uint8Array possible values
+
+/**
+ * Record played notes
+ * @param enableCanvas  Enable Canvans
+ * @param trackDuration Track duration (in seconds)
+ * @param decibelMin    Minimum binary value of decibel to track (between 0 and 255)
```
Comment (MisterPeModder) [inline front/utilities/sound/microphone.ts:25]: Those parameters don't exist anymore.

**Review by MisterPeModder — CHANGES_REQUESTED** [review]
It works! Juste a tiny comment to fix though

**Review by MisterPeModder — APPROVED** [review]

**Review by angemarchand — APPROVED** [review]

**Pull request merged**